### ISSUE TITLE
FITS helpers fail to close files

### ISSUE THREAD

**Issue by anntzer** [astropy/astropy#2824]
The `astropy.io.get{data,header}` helpers fail to close the file.
```
$ python -Wall -c 'from astropy.io import fits; fits.getdata("1-bf.fits")'
-c:1: ResourceWarning: unclosed file <_io.FileIO name='1-bf.fits' mode='rb'>
```
Not a big deal though.

**Comment by pllim** [comment]
Fixed in #4786.

**Comment by olaurino** [comment]
I am getting a similar issue with astropy v1.1.2 on Python 3.5 (continuum) when running the following command:

`python -Wall -c 'from astropy.io import fits; fits.open("data.txt")'`

Where `data.txt` is a text file.

Below is the stacktrace and the warning. I am not sure whether this should have been fixed by #4786, probably not. I can open a new issue if necessary.

````
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/export/miniconda/miniconda/envs/sherpa3/lib/python3.5/site-packages/astropy/io/fits/hdu/hdulist.py", line 138, in fitsopen
    return HDUList.fromfile(name, mode, memmap, save_backup, cache, **kwargs)
  File "/export/miniconda/miniconda/envs/sherpa3/lib/python3.5/site-packages/astropy/io/fits/hdu/hdulist.py", line 280, in fromfile
    save_backup=save_backup, cache=cache, **kwargs)
  File "/export/miniconda/miniconda/envs/sherpa3/lib/python3.5/site-packages/astropy/io/fits/hdu/hdulist.py", line 838, in _readfrom
    hdu = _BaseHDU.readfrom(ffo, **kwargs)
  File "/export/miniconda/miniconda/envs/sherpa3/lib/python3.5/site-packages/astropy/io/fits/hdu/base.py", line 423, in readfrom
    **kwargs)
  File "/export/miniconda/miniconda/envs/sherpa3/lib/python3.5/site-packages/astropy/io/fits/hdu/base.py", line 483, in _readfrom_internal
    header = Header.fromfile(data, endcard=not ignore_missing_end)
  File "/export/miniconda/miniconda/envs/sherpa3/lib/python3.5/site-packages/astropy/io/fits/header.py", line 460, in fromfile
    padding)[1]
  File "/export/miniconda/miniconda/envs/sherpa3/lib/python3.5/site-packages/astropy/io/fits/header.py", line 529, in _from_blocks
    raise IOError('Header missing END card.')
OSError: Header missing END card.
sys:1: ResourceWarning: unclosed file <_io.FileIO name='data.txt' mode='rb' closefd=True>
````
